In [61]:
import pandas as pd
import numpy as np
import ETL_functions as etf
import ast

In [62]:
df_sg = pd.read_json(r"../Datasets/steam_games.json/output_steam_games.json" ,lines=True)

In [63]:
df_sg.tail(5)

publisher                                 genres  \
120440  Ghost_RUS Games  [Casual, Indie, Simulation, Strategy]   
120441           Sacada              [Casual, Indie, Strategy]   
120442     Laush Studio            [Indie, Racing, Simulation]   
120443         SIXNAILS                        [Casual, Indie]   
120444             None                                   None   

                        app_name                     title  \
120440            Colony On Mars            Colony On Mars   
120441  LOGistICAL: South Africa  LOGistICAL: South Africa   
120442             Russian Roads             Russian Roads   
120443       EXIT 2 - Directions       EXIT 2 - Directions   
120444               Maze Run VR                      None   

                                                      url release_date  \
120440  http://store.steampowered.com/app/773640/Colon...   2018-01-04   
120441  http://store.steampowered.com/app/733530/LOGis...   2018-01-04   
120442  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
120443  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
120444  http://store.steampowered.com/app/681550/Maze_...         None   

                                                     tags  \
120440              [Strategy, Indie, Casual, Simulation]   
120441                          [Strategy, Indie, Casual]   
120442                        [Indie, Simulation, Racing]   
120443  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...   
120444  [Early Access, Adventure, Indie, Action, Simul...   

                                              reviews_url  \
120440  http://steamcommunity.com/app/773640/reviews/?...   
120441  http://steamcommunity.com/app/733530/reviews/?...   
120442  http://steamcommunity.com/app/610660/reviews/?...   
120443  http://steamcommunity.com/app/658870/reviews/?...   
120444  http://steamcommunity.com/app/681550/reviews/?...   

                                                    specs price  early_access  \
120440                [Single-player, Steam Achievements]  1.99           0.0   
120441  [Single-player, Steam Achievements, Steam Clou...  4.99           0.0   
120442  [Single-player, Steam Achievements, Steam Trad...  1.99           0.0   
120443   [Single-player, Steam Achievements, Steam Cloud]  4.99           0.0   
120444  [Single-player, Stats, Steam Leaderboards, HTC...  4.99           1.0   

              id                 developer  
120440  773640.0        Nikita "Ghost_RUS"  
120441  733530.0                    Sacada  
120442  610660.0  Laush Dmitriy Sergeevich  
120443  658870.0             xropi,stev3ns  
120444  681550.0                      None

    para esta funcion solo utilizaremos(en este df):
* id(para relacionar con la review)
* developer(va a ser por lo que vamos a filtrar/el dato que se pasa a la funcion)

In [64]:
del df_sg["genres"]
del df_sg["app_name"]
del df_sg["title"]
del df_sg["url"]
del df_sg["release_date"]
del df_sg["tags"]
del df_sg["reviews_url"]
del df_sg["specs"]
del df_sg["price"]
del df_sg["early_access"]
del df_sg["publisher"]

    con esto borramos todas las columnas que no vamos a usar

In [65]:
etf.data_review(df_sg)


Total rows:  120445

Total full null rows:  88311

Total duplicated rows: 88311


Column                                dType  No_Null_%  No_Null_Qty  \
0         id                    [<class 'float'>]      26.68        32133   
1  developer  [<class 'NoneType'>, <class 'str'>]      23.94        28836   

   Null_%  Null_Qty  
0   73.32     88312  
1   76.06     91609

    ahora hay que tratar duplicados y nulos

## eliminacion de duplicados

    cualquier registro duplicado no resulta importante debido a que los datos ya se encuentran en el dataset 
    (en especial en este caso que tenemos muchas columnas que tienen todas la columnas en null)

In [66]:
df_sg.drop_duplicates(inplace=True)#elimino todos los duplicados dejando solo registros unicos y con ellos una columna llena de valores null que removere en la siguiente instruccion

    al dropear duplicados nos dejo con cada registro unico incluyendo  una de las lineas con todos sus valores nulos 

In [67]:
df_sg.head(3)

id         developer
0           NaN              None
88310  761140.0         Kotoshiro
88311  643980.0  Secret Level SRL

    como en este caso quedo en la primera linea la dropeare indicando el indice

In [68]:
df_sg.drop(0,inplace=True)#borramos la unica columna que no poseia datos que quedaba

In [69]:
df_sg.head()

id         developer
88310  761140.0         Kotoshiro
88311  643980.0  Secret Level SRL
88312  670290.0      Poolians.com
88313  767400.0              彼岸领域
88314  773570.0              None

In [70]:
etf.data_review(df_sg)


Total rows:  32133

Total full null rows:  0

Total duplicated rows: 0


Column                                dType  No_Null_%  No_Null_Qty  \
0         id                    [<class 'float'>]     100.00        32132   
1  developer  [<class 'str'>, <class 'NoneType'>]      89.74        28835   

   Null_%  Null_Qty  
0    0.00         1  
1   10.26      3298

    ahora ya no tenemos filas completas con nulos o filas duplicadas lo que no quiere decir que no haya nulos
    como podemos comprobar en la ultima columna de el retorno de etf.data_review donde nos dice
|columna|cantidad de nulos|
|-------|-----------------|
|id| 1
|developer| 3298

## eliminacion  de nulos

    borrare cualquier nulo en developer

In [71]:
mask = df_sg["developer"].isna() #

In [72]:
df_sg = df_sg[~mask] # uso ~ para invertir la condicion de el  filtro lo que traera todos los que developer no es nulo

    ahora tratare el nulo de id

In [73]:
df_sg.isna().sum()

id           1
developer    0
dtype: int64

In [74]:
df_sg.dropna(inplace=True)

In [75]:
etf.data_review(df_sg)


Total rows:  28834

Total full null rows:  0

Total duplicated rows: 0


Column              dType  No_Null_%  No_Null_Qty  Null_%  Null_Qty
0         id  [<class 'float'>]      100.0        28834     0.0         0
1  developer    [<class 'str'>]      100.0        28834     0.0         0

    ya contamos  con este dataframe completamente libre de duplicados o nulos

In [79]:
df_sg

id                 developer
88310   761140                 Kotoshiro
88311   643980          Secret Level SRL
88312   670290              Poolians.com
88313   767400                      彼岸领域
88315   772540       Trickjump Games Ltd
...        ...                       ...
120439  745400           Bidoniera Games
120440  773640        Nikita "Ghost_RUS"
120441  733530                    Sacada
120442  610660  Laush Dmitriy Sergeevich
120443  658870             xropi,stev3ns

[28834 rows x 2 columns]

    paso id a int para sacarles la parte decimal

In [78]:
df_sg["id"] = df_sg["id"].astype(int)

    paso id a str para realizar  el merge

In [80]:
df_sg["id"] = df_sg["id"].astype(str)

In [81]:
df_sg

id                 developer
88310   761140                 Kotoshiro
88311   643980          Secret Level SRL
88312   670290              Poolians.com
88313   767400                      彼岸领域
88315   772540       Trickjump Games Ltd
...        ...                       ...
120439  745400           Bidoniera Games
120440  773640        Nikita "Ghost_RUS"
120441  733530                    Sacada
120442  610660  Laush Dmitriy Sergeevich
120443  658870             xropi,stev3ns

[28834 rows x 2 columns]

<hr>

In [82]:
df_rd = pd.read_parquet("../app/reviews_dict.parquet")

In [83]:
df_rd

posted item_id  recommend  \
0    Posted November 5, 2011.    1250       True   
1       Posted July 15, 2011.   22200       True   
2      Posted April 21, 2011.   43110       True   
0       Posted June 24, 2014.  251610       True   
1   Posted September 8, 2013.  227300       True   
..                        ...     ...        ...   
2             Posted July 10.      70       True   
3              Posted July 8.  362890       True   
0              Posted July 3.  273110       True   
1             Posted July 20.     730       True   
2              Posted July 2.     440       True   

                                               review  sentimiento  
0   Simple yet with great replayability. In my opi...            2  
1                It's unique and worth a playthrough.            2  
2   Great atmosphere. The gunplay can be a bit chu...            2  
0   I know what you think when you see this title ...            2  
1   For a simple (it's actually not all that simpl...            0  
..                                                ...          ...  
2   a must have classic from steam definitely wort...            2  
3   this game is a perfect remake of the original ...            2  
0   had so much fun plaing this and collecting res...            2  
1                                                  :D            2  
2                                      so much fun :D            2  

[58424 rows x 5 columns]

    de este dataframe usaremos todas las columnas menos review
* posted(para sacar el año)
* item_id(para relacionar los dataframe)
* recommend(cantidades)
* sentimiento(cantidades)

In [84]:
del df_rd["review"]

In [85]:
df_rd

posted item_id  recommend  sentimiento
0    Posted November 5, 2011.    1250       True            2
1       Posted July 15, 2011.   22200       True            2
2      Posted April 21, 2011.   43110       True            2
0       Posted June 24, 2014.  251610       True            2
1   Posted September 8, 2013.  227300       True            0
..                        ...     ...        ...          ...
2             Posted July 10.      70       True            2
3              Posted July 8.  362890       True            2
0              Posted July 3.  273110       True            2
1             Posted July 20.     730       True            2
2              Posted July 2.     440       True            2

[58424 rows x 4 columns]

<hr>

In [ ]:
@app.get("/UsersWorstDeveloper")
async def UsersWorstDeveloper( año : int ): #Devuelve el top 3 de desarrolladoras con juegos MENOS recomendados por usuarios para el año dado. (reviews.recommend = False y comentarios negativos)
    return {} #Ejemplo de retorno: [{"Puesto 1" : X}, {"Puesto 2" : Y},{"Puesto 3" : Z}]

<hr>

In [86]:
merge = df_rd.merge(df_sg,how="inner",left_on="item_id",right_on="id")

In [87]:
merge

posted item_id  recommend  sentimiento      id  \
0      Posted November 5, 2011.    1250       True            2    1250   
1        Posted March 30, 2015.    1250       True            2    1250   
2         Posted July 12, 2013.    1250       True            2    1250   
3       Posted August 13, 2015.    1250       True            1    1250   
4         Posted April 5, 2014.    1250       True            1    1250   
...                         ...     ...        ...          ...     ...   
49614  Posted October 29, 2015.  307130       True            2  307130   
49615             Posted May 5.  209120       True            0  209120   
49616         Posted August 13.  220090       True            0  220090   
49617         Posted August 13.  262850       True            0  262850   
49618           Posted March 5.  431510       True            2  431510   

                  developer  
0      Tripwire Interactive  
1      Tripwire Interactive  
2      Tripwire Interactive  
3      Tripwire Interactive  
4      Tripwire Interactive  
...                     ...  
49614         Legend Studio  
49615   Capcom U.S.A., Inc.  
49616             SkyGoblin  
49617             SkyGoblin  
49618    Aeon Dream Studios  

[49619 rows x 6 columns]

In [99]:
merge.to_parquet(r"../app/df_recommend.parquet")

In [ ]:
anio = 2015

In [94]:
def usersworstdeveloper(anio):
    df_f = merge[merge['posted'].str.contains(str(anio))] #filtro el df al año
    df_f = df_f[(df_f['sentimiento'] == 0) | (df_f['recommend'] == False)]#filtro el df a los que al menos poseen uno de las 2 cosas que vamos a contar

    conteo_sentimiento = df_f[df_f['sentimiento'] == 2].groupby('item_id')['sentimiento'].count()
    conteo_recommend = df_f[df_f['recommend'] == False].groupby('item_id')['recommend'].count()

    conteo_df = pd.merge( conteo_recommend,conteo_sentimiento, on='item_id', how='inner')

    df_f = df_f.merge(conteo_df,how="inner",left_on="id",right_on="item_id")

    df_f = df_f.drop_duplicates(subset=['item_id'])

    df_f = df_f.sort_values(by=['recommend_y','sentimiento_y'], ascending=[False, False])

    return df_f


In [96]:
usersworstdeveloper(2015)

posted item_id  recommend_x  sentimiento_x      id  \
21    Posted November 23, 2015.     730         True              0     730   
860   Posted November 26, 2015.  221100        False              0  221100   
1206     Posted August 1, 2015.  252490         True              0  252490   
1805   Posted October 17, 2015.  271590        False              0  271590   
1604   Posted October 16, 2015.  346110         True              0  346110   
...                         ...     ...          ...            ...     ...   
2973  Posted December 29, 2015.  359310        False              2  359310   
2974      Posted June 11, 2015.  373460        False              2  373460   
2975   Posted January 24, 2015.  233310        False              2  233310   
2976  Posted November 27, 2015.  344820        False              2  344820   
2977      Posted April 4, 2015.   27940        False              2   27940   

                                              developer  recommend_y  \
21                                                Valve          129   
860                                 Bohemia Interactive           72   
1206                                  Facepunch Studios           47   
1805                                     Rockstar North           41   
1604  Studio Wildcard,Instinct Games,Efecto Studios,...           36   
...                                                 ...          ...   
2973                                        Shiro Games            1   
2974                                     Spectral Games            1   
2975                                 Spiderweb Software            1   
2976                                       Photon Arena            1   
2977                                    DnS Development            1   

      sentimiento_y  
21               39  
860              18  
1206             10  
1805              8  
1604              7  
...             ...  
2973              1  
2974              1  
2975              1  
2976              1  
2977              1  

[315 rows x 8 columns]

<hr>

In [97]:
def usersbestdeveloper(anio):
    df_f = merge[merge['posted'].str.contains(str(anio))] #filtro el df al año
    df_f = df_f[(df_f['sentimiento'] == 2) | (df_f['recommend'] == True)]#filtro el df a los que al menos poseen uno de las 2 cosas que vamos a contar

    conteo_sentimiento = df_f[df_f['sentimiento'] == 2].groupby('item_id')['sentimiento'].count()
    conteo_recommend = df_f[df_f['recommend'] == False].groupby('item_id')['recommend'].count()

    conteo_df = pd.merge( conteo_recommend,conteo_sentimiento, on='item_id', how='inner')

    df_f = df_f.merge(conteo_df,how="inner",left_on="id",right_on="item_id")

    df_f = df_f.drop_duplicates(subset=['item_id'])

    df_f = df_f.sort_values(by=['recommend_y','sentimiento_y'], ascending=[False, False])

    return df_f

In [98]:
usersbestdeveloper(2015)

posted item_id  recommend_x  sentimiento_x      id  \
91        Posted June 19, 2015.     730         True              2     730   
3458  Posted November 12, 2015.  221100         True              2  221100   
4740      Posted June 26, 2015.  333930         True              0  333930   
5186  Posted November 18, 2015.  227940         True              0  227940   
1849     Posted March 12, 2015.     440         True              2     440   
...                         ...     ...          ...            ...     ...   
8477  Posted December 29, 2015.  359310        False              2  359310   
8478      Posted June 11, 2015.  373460        False              2  373460   
8479   Posted January 24, 2015.  233310        False              2  233310   
8480  Posted November 27, 2015.  344820        False              2  344820   
8481      Posted April 4, 2015.   27940        False              2   27940   

                developer  recommend_y  sentimiento_y  
91                  Valve           39            869  
3458  Bohemia Interactive           18             67  
4740        Splash Damage           13             75  
5186            Reto-Moto           11             62  
1849                Valve           10            391  
...                   ...          ...            ...  
8477          Shiro Games            1              1  
8478       Spectral Games            1              1  
8479   Spiderweb Software            1              1  
8480         Photon Arena            1              1  
8481      DnS Development            1              1  

[315 rows x 8 columns]